src/test/fifotest.cpp

```diff
@@ -0,0 +1,182 @@
+#include <gtest/gtest.h>
+
+#include <algorithm>
+#include <random>
+
+#include "util/fifo.h"
+
+namespace {
+
+class FifoTest : public testing::Test {
+};
+
+TEST_F(FifoTest, writeAvailableTest) {
+    std::vector<float> data(1024);
+    FIFO<float> fifo(1024);
+    ASSERT_EQ(1024, fifo.writeAvailable());
+    ASSERT_EQ(100, fifo.write(data.data(), 100));
+    ASSERT_EQ(924, fifo.writeAvailable());
+    ASSERT_EQ(50, fifo.read(data.data(), 50));
+    ASSERT_EQ(974, fifo.writeAvailable());
+    ASSERT_EQ(974, fifo.write(data.data(), 10000));
+    ASSERT_EQ(0, fifo.writeAvailable());
+}
+
+TEST_F(FifoTest, readAvailableTest) {
+    std::vector<float> data(1024);
+    FIFO<float> fifo(1024);
```

Comment:
a (potentially Value-Parameterized) test that tests a couple more (non power-of-two) queue sizes would make sense, wdyt?

Reply:
Yes, also the offset near the numeric limit of the indices should be parametrized

Reply:
Note though that internally the queue sizes are always rounded up to the next power of two. Still, good to add parametrized tests.

Reply:
right, but if the size and the rounded up size are always the same and the implementation mistakenly uses the wrong values, the edge case is not caught. So even though it may not matter with the current implementation we should add a unit test to ensure this mistake is not made accidentally in the future.

Reply:
Sure, sure, and once parametrized, we can add whatever size.

Reply:
done. please check and resolve.